### ISSUE TITLE
No warning/error message when from_dict_of_lists is passed a dictionary with string type keys

### ISSUE THREAD

**Issue by snu-ceyda** [networkx/networkx#2493]
**Version**: networkx (1.11)

We were using the following code to generate the adjacency matrix from a dictionary representation
```
dict={0: [1], 1: [2], 2: [3]} #some graph
nx.adjacency_matrix(nx.from_dict_of_lists(dict))
```
We found out that on some places the adjacency matrix generated was wrong and after hours of digging ...
we figured out that somewhere in the process the dictionary was json encode/decoded which resulted in a dictionary with string typed keys . `dict={"0": [1], "1": [2], "2": [3]}`
which changes the output of `nx.from_dict_of_lists`. I don't know networkx library that well but if it is intended behavior it should be explained in the [docs](http://networkx.readthedocs.io/en/stable/reference/generated/networkx.convert.from_dict_of_lists.html?). or maybe should log a warning message.

Now you can just say we should have been more careful but I think it is an easy enough mistake to make. Hope this helps someone

**TLDR**; when dictionary keys are strings it results in a different graph, which can effect other operations
```
import networkx as nx 
test_string={"0": [1], "1": [2], "2": [3]}
wrong=nx.adjacency_matrix(nx.from_dict_of_lists(test_string))
test_int={0: [1], 1: [2], 2: [3]}
true=nx.adjacency_matrix(nx.from_dict_of_lists(test_int))
print(wrong)
print(true)
```

**Comment by dschult** [comment]
It looks like the conversion from integers to strings is the culprit here.  Note that if the conversion had been made in both the keys and the list of neighbors like ```{"0": ["1"], ...``` then the graph would have been as expected.  

Are you requesting that "0" be treated like 0 when creating nodes?  Or maybe you are suggesting a doc change...  what should it say and where should it be? I think I'm not quite clear yet on how to improve this.

**Comment by snu-ceyda** [comment]
Yes, I see now that when `{"0": ["1"], ...` it is also works correctly. It seems to be wrong when type(dictionary.key)  is not the same as the type of elements in array(i.e dictionary.value). 
For example  `{0: ["1"], 1: ["2"], 2: ["3"]}` and `{0: ["1"], "1": ["2"], "2": ["3"]}` etc also produces wrong results. 

So maybe throwing an error when there is a type mismatch might be good? Otherwise mistakes of this type currently may go unnoticed.
Or would casting from string to int(or else) be better?
Thank you for the quick reply

**Comment by dschult** [comment]
The problem is that we can't know ahead of time whether the user is looking for a string node or an int node. And some graphs may have both string nodes and int nodes in the same graph. 

Your results are interesting. The first two are, of course isomorphic to each other as they should be. The order of the nodes differs but that is not surprising. The combination graphs have more nodes because strings are not equal to ints and so we have nodes: ```0,1,2,3,"0","1","2","3"```.

Do you have any suggestions for how to allow people to use whatever nodes they like while helping them find issues like you found?
